Import TWIR (#592)
rails/website: _posts/2025-10-17-this-week-in-rails.markdown

---
layout: post
title: "Bound SQL literals in CTEs, negative scopes for enums and more!"
categories: news
author: vipulnsward
og_image: assets/images/this-week-in-rails.png
published: true
date: 2025-10-17
---

Hi, it's [Vipul!](https://www.saeloun.com/team/vipul/). Let's explore this week's changes in the Rails codebase.

[Add missing structured event data for logs](https://github.com/rails/rails/pull/55922)  

> Followup on https://github.com/rails/rails/pull/55904. 

Reviewer: ```suggestion
Follow-up on <https://github.com/rails/rails/pull/55904>. 
```

> [Add support for bound SQL literals in CTEs](https://github.com/rails/rails/pull/55918)  

Reviewer: Is this not private API? I'm not sure we should include it

This Pull Request updates the method to accept both `Arel::Nodes::SqlLiteral` and `Arel::Nodes::BoundSqlLiteral`.

[Fix passing both module: and shallow: to resources](https://github.com/rails/rails/pull/55916)  
This broke because the keyword arg change started always passing `shallow:` through to this `scope` block. 
However, this caused an issue  because passing the `shallow` option caused it to _always_ override the  scope's current `shallow` value, instead of only doing that when specified.
This commit fixes the behavior by only passing `shallow` to the `scope` when it is specified, instead of unconditionally.

[Add advisory DB update by default in bin/bundler-audit](https://github.com/rails/rails/pull/55913)  
This Pull Request updates the `bin/bundler-audit` script to include the `--update` option.

[Fix negative scopes for enums to include records with `nil` values](https://github.com/rails/rails/pull/55912)    
Since NULL != NULL in SQL, the condition where `column != 'value'` always ignores rows with NULLs, but should include them too.

> [Show help hint when starting `bin/rails console`](https://github.com/rails/rails/pull/55910)  

Reviewer: @claudiob @gregmolnar I honestly don't know how I feel about this anymore, we avoid using fixed-width in titles because of this one CSS rule, but I feel the CSS is the problem... not us :joy:

When altering a table in SQLite3 that is referenced by child tables with `ON DELETE CASCADE` foreign keys, ActiveRecord would silently delete all data from the child tables. This occurred because SQLite requires table recreation for schema changes, and during this process:

1. The original table is temporarily dropped
2. `CASCADE` deletes fire immediately on child tables
3. The table is recreated, but child data is already gone

The fix simply reverses the order to follow the [official SQLite 12-step ALTER TABLE procedure](https://www.sqlite.org/lang_altertable.html#otheralter).

[Make the Structured Event Subscriber emit events in format that are useful for logging](https://github.com/rails/rails/pull/55904

[Improve support for SQLite database URIs](https://github.com/rails/rails/pull/55896)  
This commit introduces a new singleton method on the connection adapter, `SQLite3Adapter.resolve_path`, to reliably resolve a filesystem path from the database configuration.

It uses this method in the connection adapter's initializer to ensure the database file's directory exists. Previously the adapter omitted this check for database URIs.

> It also introduces usage of this method in the `create` and `drop` database tasks which also previously did not fully support database URIs.

Reviewer: ```suggestion
This PR also adds usage of this method in the `create` and `drop` database tasks which previously did not fully support database URIs.
```

[Fix verbose redirect output and route source mapping](https://github.com/rails/rails/pull/55889)  
This Pull Request changes mostly dispatch code to record some source location in redirect endpoints when needed. This allows this feature to work in environments other than development, if needed, instead of depending on the other route source locations feature.

This adds a parameter to `redirect.action_dispatch` which may not be desirable. We can instead hack the caller trace to include the source location or use route source locations. Open to feedback on this, but I prefer to keep things simple here with an event parameter.

> [Load core extensions with require_relative](https://github.com/rails/rails/pull/55887)    

Reviewer: ```suggestion
[Load core extensions with require_relative](https://github.com/rails/rails/pull/55887)  
```

`require_relative` is more aligned with this very intentional file structure, and we get some less `$LOAD_PATH` lookups as an extra ball.

[Only preconnect connections that have been prepopulated](https://github.com/rails/rails/pull/55857)  
Otherwise-manually-created connections that have not directly cause a connection should remain lazy.

[Refactor gcs_service public interface to bring more in line with s3_service](https://github.com/rails/rails/pull/55856)  

> The `s3_service` exposes it's respective client and bucket attributes publicly. I think this makes sense in the context of active storage, as it allows users to rely on the same adapter & configs that the framework is using for Activestorage without having to duplicate your own client and ensure it matches the original, which can be modified at runtime, in order to perform operations outside of the scope of ActiveStorage.

Reviewer 2: Replace `it's` with `its`

[Make `ActiveRecord::Assertions::QueryAssertions` method outputs consistent](https://github.com/rails/rails/pull/55852)  

[Skip analyze job if ruby-vips or mini_magick gem are missing](https://github.com/rails/rails/pull/55839)

[Add structured event for Rails deprecations, when `config.active_support.deprecation` is set to `:notify`.](https://github.com/rails/rails/pull/55838)  
This change makes it dependent of the `:notify` behavior than adding yet another behavior just for structured events

[Fix `change_column` to preserve old column attributes for sqlite3](https://github.com/rails/rails/pull/55837)  

[Add prefix option to has_secure_token for improved token identification](https://github.com/rails/rails/pull/55822)  

> Adds an optional `:prefix` parameter to has_secure_token that prepends a string to generated tokens, making token types immediately identifiable in logs, debugging sessions, and error messages

Reviewer: ```suggestion
Adds an optional `:prefix` parameter to `has_secure_token` that prepends a string to generated tokens, making token types immediately identifiable in logs, debugging sessions, and error messages
```
